### PR TITLE
Enhance per bucket, and per document monitor notification message ctx.

### PR DESCRIPTION
**Pull request by AWSHurneyt**
*Issue #, if available:*
https://github.com/opensearch-project/alerting/issues/1300
https://github.com/opensearch-project/alerting/issues/1396
https://github.com/opensearch-project/alerting/issues/1401

*Description of changes:*
1. Added support for returning sample documents for bucket level monitors, and document level monitors.
2. Added support for printing query/rule info in notification messages for document level monitors.

*CheckList:*
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/alerting/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by eirsep** [comment]
typically we create smaller PRs. one for each issue is a practice we should try follow

**Comment by AWSHurneyt** [comment]
The nonsecurity-related test failures seem related to the fixes in PR https://github.com/opensearch-project/alerting/pull/1464

The security-related test failures have been failing for some time, so they're unrelated to these changes.
https://github.com/opensearch-project/alerting/actions/workflows/security-test-workflow.yml?query=branch%3Amain

**Comment by AWSHurneyt** [comment]
> typically we create smaller PRs. one for each issue is a practice we should try follow

@eirsep agreed. Would you say that's blocking for this PR? All of the changes will require the AlertContext class, so the second PR couldn't be raised until the first is merged.

File: alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt [hunk 1]
```diff
@@ -230,6 +235,33 @@ object BucketLevelMonitorRunner : MonitorRunner() {
...
+                    val searchRequest = monitorCtx.inputService!!.getSearchRequest(
+                        monitor = monitor.copy(triggers = listOf(trigger)),
+                        searchInput = monitor.inputs[0] as SearchInput,
+                        periodStart = periodStart,
+                        periodEnd = periodEnd,
+                        prevResult = monitorResult.inputResults,
+                        matchingDocIdsPerIndex = null,
+                        returnSampleDocs = true
+                    )
+                    monitorCtx.client
```
Comment (AWSHurneyt) [inline alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt:252]: This line is dev code that should be removed. I'll be pushing a revision shortly.

File: alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt [hunk 2]
```diff
@@ -220,6 +224,7 @@ object BucketLevelMonitorRunner : MonitorRunner() {
                     ?.addAll(monitorCtx.alertService!!.convertToCompletedAlerts(keysToAlertsMap))
         }
```
Comment (eirsep) [inline alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt:226]: What are the keys of this map variable and the inner map ?
Can we add some code comments to elaborate about how the docs are held in memory

File: alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt [hunk 3]
```diff
@@ -480,17 +516,88 @@ object BucketLevelMonitorRunner : MonitorRunner() {
...
+    private fun getAlertContext(
+        alert: Alert,
+        alertSampleDocs: Map<String, Map<String, List<Map<String, Any>>>>
+    ): AlertContext {
+        val bucketKey = alert.aggregationResultBucket?.getBucketKeysHash()
+        val sampleDocs = alertSampleDocs[alert.triggerId]?.get(bucketKey)
+        return if (!bucketKey.isNullOrEmpty() && !sampleDocs.isNullOrEmpty()) {
+            AlertContext(alert = alert, sampleDocs = sampleDocs)
+        } else {
+            logger.warn(
```
Comment (eirsep) [inline alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt:542]: Error log

File: alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt [hunk 4]
```diff
@@ -480,17 +516,88 @@ object BucketLevelMonitorRunner : MonitorRunner() {
...
+        client: Client,
+        monitorId: String,
+        triggerId: String,
+        searchRequest: SearchRequest
+    ): Map<String, List<Map<String, Any>>> {
+        val sampleDocumentsByBucket = mutableMapOf<String, List<Map<String, Any>>>()
+
+        val searchResponse: SearchResponse = client.suspendUntil { client.search(searchRequest, it) }
+        val aggs = searchResponse.convertToMap().getOrDefault("aggregations", mapOf<String, Any>()) as Map<String, Any>
+        val compositeAgg = aggs.getOrDefault("composite_agg", mapOf<String, Any>()) as Map<String, Any>
```
Comment (eirsep) [inline alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt:564]: Are all aggs of composite type??

File: alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt [hunk 5]
```diff
@@ -480,17 +516,88 @@ object BucketLevelMonitorRunner : MonitorRunner() {
...
+                alert.triggerId,
+                alert.monitorId,
+                alert.executionId
+            )
+            AlertContext(alert = alert, sampleDocs = listOf())
+        }
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    private suspend fun getSampleDocs(
```
Comment (eirsep) [inline alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt:554]: Javadocs

**Review by eirsep — CHANGES_REQUESTED** [review]
Skimmed through the PR. Need to take a deeper look to understand the top hits and low hits logic

File: alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt (hunk 2 above)
Comment (AWSHurneyt) [reply alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt:226]: Good callout. I've added a comment to explain that the `alertSampleDocs` map is `Map<TriggerId, Map<BucketKeysHash, List<Alert>>>`.

File: alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt (hunk 3 above)
Comment (AWSHurneyt) [reply alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt:542]: Revised.

File: alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt (hunk 5 above)
Comment (AWSHurneyt) [reply alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt:561]: Added a comment with details.

File: alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt (hunk 4 above)
Comment (AWSHurneyt) [reply alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt:570]: The frontend visual editor translates the bucket monitor creation form selections into a composite aggregation ([source](https://github.com/opensearch-project/alerting-dashboards-plugin/blob/3b6a9bfc0c8f1b5a1df393714aea695b399ace5c/public/pages/CreateMonitor/containers/CreateMonitor/utils/formikToMonitor.js#L239)); however, different aggregation types could potentially be configured using the extraction query editor. 

As discussed offline, I'll take a follow-up item to support more than composite aggregations.

Comment (AWSHurneyt) [reply alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt:570]: Created issue https://github.com/opensearch-project/alerting/issues/1473 to track this follow-up item.

File: alerting/src/main/kotlin/org/opensearch/alerting/script/BucketLevelTriggerExecutionContext.kt [hunk 6]
```diff
@@ -5,21 +5,25 @@
...
+
 data class BucketLevelTriggerExecutionContext(
     override val monitor: Monitor,
     val trigger: BucketLevelTrigger,
     override val results: List<Map<String, Any>>,
     override val periodStart: Instant,
     override val periodEnd: Instant,
     val dedupedAlerts: List<Alert> = listOf(),
-    val newAlerts: List<Alert> = listOf(),
+    val newAlerts: List<AlertContext> = listOf(),
```
Comment (eirsep) [inline alerting/src/main/kotlin/org/opensearch/alerting/script/BucketLevelTriggerExecutionContext.kt:26]: could have been a separate field to have alert context list and retain new alerts

File: alerting/src/main/kotlin/org/opensearch/alerting/util/AlertingUtils.kt [hunk 7]
```diff
@@ -177,3 +186,85 @@ fun ThreadContext.StoredContext.closeFinally(cause: Throwable?) = when (cause) {
...
+/**
+ * Mustache template supports iterating through a list using a `{{#listVariable}}{{/listVariable}}` block.
+ * https://mustache.github.io/mustache.5.html
+ *
+ * This function looks for `{{#${[AlertContext.SAMPLE_DOCS_FIELD]}}}{{/${[AlertContext.SAMPLE_DOCS_FIELD]}}}` blocks,
+ * and parses the contents for tags, which we interpret as fields within the sample document.
+ *
+ * @return a [Set] of [String]s indicating fields within a document.
+ */
+fun parseSampleDocTags(messageTemplate: Script): Set<String> {
```
Comment (eirsep) [inline alerting/src/main/kotlin/org/opensearch/alerting/util/AlertingUtils.kt:199]: we should ideally have a parameter `enrichContextWithSampleDocs` in the create monitor request inside the trigger payload

File: alerting/src/test/kotlin/org/opensearch/alerting/DocumentMonitorRunnerIT.kt [hunk 8]
```diff
@@ -2428,6 +2428,222 @@ class DocumentMonitorRunnerIT : AlertingRestTestCase() {
...
+                    assertTrue(
+                        "The notification message is missing the query name.",
+                        actionOutput["message"]!!.contains("(name=${docQuery.name})")
+                    )
+                }
+            }
+        }
+    }
+
+    fun `test expected document and rules print in notification message`() {
```
Comment (eirsep) [inline alerting/src/test/kotlin/org/opensearch/alerting/DocumentMonitorRunnerIT.kt:2493]: can we add a test for scenario where there is error alert from the trigger evaluation

File: alerting/src/test/kotlin/org/opensearch/alerting/MonitorRunnerServiceIT.kt [hunk 9]
```diff
@@ -1989,6 +1990,88 @@ class MonitorRunnerServiceIT : AlertingRestTestCase() {
         }
     }
 
+    fun `test bucket-level monitor notification message includes sample docs per bucket`() {
```
Comment (eirsep) [inline alerting/src/test/kotlin/org/opensearch/alerting/MonitorRunnerServiceIT.kt:1993]: plz add more test coverage . scenarios verifying each of the checks to printsSampleDocs

**Review by eirsep — APPROVED** [review]

File: alerting/src/main/kotlin/org/opensearch/alerting/script/BucketLevelTriggerExecutionContext.kt (hunk 6 above)
Comment (AWSHurneyt) [reply alerting/src/main/kotlin/org/opensearch/alerting/script/BucketLevelTriggerExecutionContext.kt:26]: I'll add this as a follow-up item.

File: alerting/src/test/kotlin/org/opensearch/alerting/DocumentMonitorRunnerIT.kt (hunk 8 above)
Comment (AWSHurneyt) [reply alerting/src/test/kotlin/org/opensearch/alerting/DocumentMonitorRunnerIT.kt:2493]: I'll add this as a follow-up item.

File: alerting/src/test/kotlin/org/opensearch/alerting/MonitorRunnerServiceIT.kt (hunk 9 above)
Comment (AWSHurneyt) [reply alerting/src/test/kotlin/org/opensearch/alerting/MonitorRunnerServiceIT.kt:1993]: I'll add this as a follow-up item.

File: alerting/src/main/kotlin/org/opensearch/alerting/util/AlertingUtils.kt (hunk 7 above)
Comment (AWSHurneyt) [reply alerting/src/main/kotlin/org/opensearch/alerting/util/AlertingUtils.kt:199]: Agreed. I'll create an issue to track updating the trigger data model to add this boolean-based logic.

Comment (AWSHurneyt) [reply alerting/src/main/kotlin/org/opensearch/alerting/util/AlertingUtils.kt:199]: Created issue https://github.com/opensearch-project/alerting/issues/1475 to track this action item.

File: alerting/src/test/kotlin/org/opensearch/alerting/MonitorRunnerServiceIT.kt (hunk 9 above)
Comment (AWSHurneyt) [reply alerting/src/test/kotlin/org/opensearch/alerting/MonitorRunnerServiceIT.kt:1993]: Create issue https://github.com/opensearch-project/alerting/issues/1476 to track this follow-up item. For context, the `printsSampleDocs` function does have unit testing ([link](https://github.com/opensearch-project/alerting/pull/1450/files#diff-fb6dad353d75750210919e12bc1019bbfbef87393fee8e54e7642205e75f04a1R55)); but integration testing will increase confidence.

File: alerting/src/test/kotlin/org/opensearch/alerting/DocumentMonitorRunnerIT.kt (hunk 8 above)
Comment (AWSHurneyt) [reply alerting/src/test/kotlin/org/opensearch/alerting/DocumentMonitorRunnerIT.kt:2493]: Create issue https://github.com/opensearch-project/alerting/issues/1476 to track this follow-up item.

**Review by riysaxen-amzn — APPROVED** [review]

File: alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt [hunk 10]
```diff
@@ -230,6 +236,32 @@ object BucketLevelMonitorRunner : MonitorRunner() {
                 ?: mutableListOf()
             // Update nextAlerts so the filtered DEDUPED Alerts are reflected for PER_ALERT Action execution
             nextAlerts[trigger.id]?.set(AlertCategory.DEDUPED, dedupedAlerts)
+
+            // Only collect sample docs for triggered triggers, and only when at least 1 action prints sample doc data.
+            val isTriggered = !nextAlerts[trigger.id]?.get(AlertCategory.NEW).isNullOrEmpty()
+            if (isTriggered && printsSampleDocData(trigger)) {
+                try {
+                    val searchRequest = monitorCtx.inputService!!.getSearchRequest(
```
Comment (AWSHurneyt) [inline alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt:244]: This current implementation executes the monitor query multiple times; the first time on line 124 to collect the data for trigger evaluation, and then subsequent searches are executed for each triggered trigger in order to collect sample documents.

Ideally, we want to collect the sample documents in the call to `collectInputResults` on line 124 so we can avoid multiple queries as that will improve performance.

Comment (AWSHurneyt) [reply alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt:244]: Created issue https://github.com/opensearch-project/alerting/issues/1481 to track this follow-up item.

**Comment by chiarch84** [comment]
Dear @AWSHurneyt it is not clear to me weather this should fix or not the fact that {{ctx.results.0.hits.hits.0._source}}  doesn't print anything. I tried OpenSearch version 2.16.0 but it seems not to work as in previous versions. Do I need to use a different syntax to make it work?

**Comment by AWSHurneyt** [comment]
Hi @chiarch84!

This is the documentation, and an example for these enhancements.
https://opensearch.org/docs/latest/observing-your-data/alerting/triggers/#mustache-template-example-1

A concern for us when making these enhancements was that the number of `hits` could easily be several thousand depending on the use case. Printing out a single `_source` from that list of thousands probably wouldn't be an issue, but mustache template supports iterating through lists, and printing values from each entry. Printing data in a notification message for thousands of hits could cause the message to exceed the size restrictions accepted by the destination webhook; causing the message to fail to send. So we added the `sample_documents` variable to the `ctx` to help manage the size of the notification message. We currently return a max of 10 sample docs, but we want to enhance that so it's a configurable cluster setting (no timeline for that just yet though).

**Comment by chiarch84** [comment]
Thank you for your answer @AWSHurneyt 
I tried your suggestion in my OpenSearch version 2.16.0 but I still get an empty message.
Here the details:
Content of the message:
``` 
{{ctx.monitor.name}} just entered alert status. 
Please investigate the issue.

  - Severity: {{ctx.trigger.severity}}
  - Period start: {{ctx.periodStart}}
  - Period end: {{ctx.periodEnd}}
  - Hits:  {{ctx.results.0.hits.total.value}}
  - Alerts Info: {{ctx.results.0.hits}}

Alerts
{{#ctx.alerts}}
    Sample documents:
    {{#sample_documents}}
        Index: {{_index}}
        Document ID: {{_id}}
       
        Exc TYPE: {{_source.bdap_exception_type}}
        -----------------
    {{/sample_documents}}
    {{#associated_queries}}
        Name: {{name}}
        Id: {{id}}
        Tags: {{tags}}
    ------------------------
    {{/associated_queries}}
{{/ctx.alerts}}
```
Previewed message obtained:
```
Monitor BDAP data download - per QUERY just entered alert status. 
Please investigate the issue.

  - Severity: 3
  - Period start: 2024-11-26T12:10:52+01:00
  - Period end: 2024-11-26T12:20:52+01:00
  - Hits:  80
  - Alerts Info: [object Object]

Alerts

```
So even though there are 80 hits, I get nothing in the details part of Alerts.
Thank you for your help!

**Comment by AWSHurneyt** [comment]
@chiarch84 Just to clarify, are you only seeing the issue in the preview of the message, or is the variable missing when an actual message is received after a monitor execution?

I just tested using the 2.16.0 docker image, and was able to successfully print sample docs in the message when my per bucket monitor executed; but it does look like the preview isn't showing the samples correctly (I created issue https://github.com/opensearch-project/alerting/issues/1731 to track that visual bug).

Also, based on the message template you shared, it looks like the monitor type is per query? Unfortunately, the sample docs variable is only available for per bucket, and per document monitors currently.
https://opensearch.org/docs/latest/observing-your-data/alerting/triggers/#:~:text=Only%20available%20with%20bucket%2D%20and%20document%2Dlevel%20monitors.

**Comment by chiarch84** [comment]
@AWSHurneyt in fact my monitor is per query, so I will try with one per bucket and let you know.
I was not seeing the result nor in the preview nor in the email, but probably because it is not the correct type of monitoring.

**Pull request merged**